data/pigui/modules/radar.lua

```diff
@@ -269,10 +354,10 @@ radar3d.draw = function(self, center)
 	end
 end
 
--- This variable needs to be outside the function in order to capture state
--- between frames. We are trying to ensure that a mouse right-click started and
--- finished inside the radar area in order to trigger the popup.
-local click_on_radar = false
+local function onTargetClicked(target)
+	-- TODO: Should ships be set as nav or combat targets?
+	Game.player:SetNavTarget(target.body)
+end
 
 -- display either the 3D or the 2D radar, show a popup on right click to select
 local function displayRadar()
```

Comment:
Ideally I'd like to split this function into multiple functions, but that would require moving a LOT of variables into the file scope. Is there another way?

What is the best practise in Lua for doing this?

Reply:
Depends - what are you trying to split out of the function? You have a few options:
1. Pass read-only parameters via function arguments, granularize function actions such that they can return "changed values" as needed. This allows one "master function" to own the local functions and pass them around. If needed, the master function can pass a table value wrapping multiple variables into sub-functions, but that's basically option 2:
2. Convert the radar display into a module (see `data/pigui/libs/module.lua`, and how it's used in e.g. the save/load window) and implement most of your drawing functions as methods on the "radar module". If you do this route, I'd strongly recommend attempting to make your "display" functions non-mutating and use the built-in `message()` facilities to handle state changes.

Reply:
I was thinking input handling and actually drawing the radar are two rather different logical tasks which should be split.. so yeah, I guess I can put the entire state of the radar module into a table and pass that around (I forgot about tables..).

It's a shame the 3d radar is partially in C++ otherwise it would be much neater to refactor this thing.  If/when we upgrade to the next version of ImGUi we could move this into Lua as well since the next version of ImGui has native ellipse drawing primitives.

I'm thinking the next step will be to actually define multiple pieces of radar equipment and then have functionality depending on what equipment is installed - range, target identification, etc.  So any refactoring here should take this into account. Or possibly require further refactoring, in which case, might as well leave this as-is for now and leave the refactoring to when the next step is done.

Thoughts?

Reply:
> Or possibly require further refactoring, in which case, might as well leave this as-is for now and leave the refactoring to when the next step is done.

I'd prefer to leave this as-is - future expansions to Sensor gameplay are going to need significant buy-in from both C++ code and game design, so it's better to leave it "easy to upgrade" rather than try to anticipate our future needs before we've even figured out what those are yet.

Reply:
Leaving as-is for now..